### ISSUE TITLE
Bug: [no-floating-promises] The rule does not detect a problem when passing a pending promise as a function parameter.

### ISSUE THREAD

**Issue by lastminutegoals** [typescript-eslint/typescript-eslint#5197]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=KYDwDg9gTgLgBAYwgOwM7wLYE8BiEI4CuyCcAvHAIapYlwAUAbpQDaHABcVyWAlFwAUoEDAEtUwADwAjfC2CVkAPnIqA3gCg4cKMBiEoyOMmAB3OEJHjg9ertQQWjYL1U7gDpzZhR2vXgDcGgC+QRqgkLCIKOhw2ABClFBEdBTUtKRMrOxcinyCwmISMnIKym6a2rr6hsZmFoXWtvaOzq5kKi1e9ABmrBL+QaEaGgD0AFTjcICg5BbUqKLIAOZUxoQY0sBQcDAQOwAWwHA9xAgwoihw46MalKaUopi4+CkI9ACMgbf3j3FYiclTh8vmNJjM5qgFstVicWD1RCx5AATOBgRoSK43O4PJ54AhA7G-BJJV7Ar6Ep4A0kUv540mfck-XEvAlMurmSxFGxdZxuHneXwuQbfHF-Kms0UmDno7keVpHDruTzOeg+PzC0FTWYAOQg8ApixWMEOx1O50uSCgujOLCwcCRemAZ1Q3DgW2E22uGmwdKBnyCxMBJGBAee+ODNMD9OFUYlRLD0a+PpZEbZUoaVgkdjlXj5OZVaqFSf+JLj8HTnKa-LzyoF6pBEymWimgBlyCFQlaKODEWHwxHAJGjHosCCUc7QtGZo7UGFmi5GMBJSgYR1QAB0zbgbd1bqt0Dg4jgm0N9sdZwHG8xIvjvuDseDDKCkZLQbeyfDb0f1+ZH-oFZl2a1jW8qqoK-iMqK95vP+U6AfKwHdIW4EBEAA&eslintrc=N4KABGBECmB2BukBcZQQpARgJwPYHcBnabZMAF2wFdoAacdGQgJgAZmBGMym+9KWLgAm0btWgMAvnyh5c5MbwaQADgENsxUikgABcgE8V0QgGNsASxXkAtCYA2F2OQD06zSUgzVGrQHlrC1xYQjI0RmhTAFs1ADUSQiDYMkh7NXITBRkMQlwqbFNoABUjUR0o4Sp7UWyoFTwAK0iFFABtSAA6F3IzYIAzCwBzDobc5IBdKW8VeypBp1C2hgx9UrNLaztCR2dIBnHvamrF1GWoYiiLMlazjHwNZNqMNXt7g1Czg7PIAEcqeRM11uUHu2EewMgiVgg2qe34X34MB+0GRPyB-AwJDwpCeUBeb1CuPCGIEVXsKSGgmwomBkk+tT0hmM6ystgcTlcghsfXsuHSTkGNnquEuxBONxJMGw2K8wOJGMglNw1NiuAsQjIfRexFxGCV1IAkgaAGIAUU12okJLp8IZq2Z5lZWx2nNwNlFVGIQiFeFFgKWkqxytlJPliNMAAtIgBrQgAYWCQgs5CS2sUdGBGEjMcIqvVACVoOR8skUDwMySs1HTLGAMr1aBqIQncu0z5SECSIA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEksArngCLKYVQB0A9A3pADTgakeLAAm9DO0joSRdAA8+gmlFESMAJVmxUjQkWlFx6AUNp4AhpgDm6PpRlzFy81FkBZcZPQBJInhYsIjseJj06CrCAGawmNDoAMLEJMgkgQGJlqlENv4AYmjoAHKW8LKh4ZFukIiYslgAbugA6pZ40AAW2IyIjJURUbTGAKLwyA5gYYM1AO7tXQPVqpDpmMjQk9PLwgBG2egAqpio0hxmK6jIu+wA2m60svJKkG4AukIAvlEyCtB6YlMlHuKwAVKCuMZAQB9dQ+MhDKDgrgcCGvCAfT5AA

### Repro Code

```typescript
export const myFooFunc = async (value: any): Promise<boolean> => {
  return new Promise((resolve) => resolve(true));
};

export const myBarFunc = async (value: any): Promise<boolean> => {
  return new Promise((resolve) => resolve(false));
};

/** ✅ Passing a number to the function */
await myFooFunc(1);
await myBarFunc(1);

/** ✅ Passing a fulfilled promise */
await myFooFunc(await myBarFunc(1));
await myBarFunc(await myFooFunc(1));
await myFooFunc(await new Promise(resolve => resolve(true)));
await myBarFunc(await new Promise(resolve => resolve(true)));

/** ✅ Not awaiting the function correctly detects an error */
myFooFunc(1);
myBarFunc(1);
myFooFunc(await myBarFunc(1));
myBarFunc(await myFooFunc(1));
myFooFunc(await new Promise(resolve => resolve(true)));
myBarFunc(await new Promise(resolve => resolve(true)));

/** 
 * ❌ Passing an unfulfilled/floating promise as a function parameter.
 * ❌ No error is being detected.
 */
await myFooFunc(myBarFunc(1));
await myBarFunc(myFooFunc(1));
await myFooFunc(new Promise(resolve => resolve(true)));
await myBarFunc(new Promise(resolve => resolve(true)));
```

### ESLint Config

```json
{
    "parser": "@typescript-eslint/parser",
    "plugins": [
        "@typescript-eslint"
    ],
    "rules": {
        "@typescript-eslint/no-floating-promises": [
            "error",
            {
                "ignoreVoid": false,
                "ignoreIIFE": false
            }
        ],
        "@typescript-eslint/no-misused-promises": [
            "error",
            {
                "checksConditionals": true,
                "checksVoidReturn": true,
                "checksSpreads": true
            }
        ]
    }
}
```

### tsconfig

```jsonc
{
    "compilerOptions": {
        "module": "esnext",
        "moduleResolution": "node",
        "target": "esnext",
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "isolatedModules": true,
        "noEmit": true,
        "watch": true,
        "strict": true,
    }
}
```

### Expected Result

I expected that passing a floating/unfulfilled/pending promise inside as a function parameter would make ESLint detect it as a problem.

```typescript
export const myFooFunc = async (value: any): Promise<boolean> => {
  return new Promise((resolve) => resolve(true));
};

await myFooFunc(new Promise(resolve => resolve(true)));
// new Promise() inside the function parameter is not fulfilled (yet)
```

### Actual Result

ESLint didn't detect a problem.

### Additional Info

I am not entirely certain whether this is a bug or the intended behavior, but it made sense to me that ESLint would detect this as an problem, so I thought I would report it.

In case it's not a bug, I would like to ask if there is a workaround or an alternative configuration that would let me achieve the same results as the one described above.

Thanks!

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.27.1` |
| `@typescript-eslint/parser`            | `5.27.1` |
| `TypeScript`                           | `4.7.3` |
| `ESLint`                               | `8.17.0` |
| `node`                                 | `16.15.0` |

**Comment by bradzacher** [comment]
This is intentional - the rule only looks for expression statements that are typed as promise and aren't properly handled.

For this case this isn't a floating promise - any lint rule that we have assumes that the function handles the promise, as it is the correct assumption in almost every case. To do otherwise would be to introduce many false-positives.

The best course of action is to use stronger types, and never use `any`.

**Comment by lastminutegoals** [comment]
Thanks for the lighting fast response.

If I understand correctly, you are saying that the rule doesn't look at what is being passed as params inside of a function, it only checks how functions are called, when assigning something to a variable/const and in similar scenarios.

Just for the record, I only used `any` in this demo example, but you can pass a `string` or an interface and you would still get the same results.

Either way, thanks again for the information, and the hint you gave about using stronger types is a good one, I suppose I can use something like `Awaited<ReturnType<typeof myBarFunc>>` as an alternative (a stronger type) if this can't be handled by ESLint alone.

**Comment by bradzacher** [comment]
> If I understand correctly, you are saying that the rule doesn't look at what is being passed as params inside of a function, it only checks how functions are called, when assigning something to a variable/const and in similar scenarios.

It does not check assignments - only top-level expression statements.
Check out the examples in the docs: https://typescript-eslint.io/rules/no-floating-promises#rule-details

---

The hint about stronger types is to secure your function inputs so that TS can catch when you're passing promises into places they may not be correctly handled.
`any`/`unknown` should be a last resort!

If your function is typed to accept promises, then the rule will check as it would any other - as long as you use the variable somewhere.
If you don't use the variable then a rule like `no-unused-vars` will error on it.

**Comment by lastminutegoals** [comment]
I've read the docs, but to be honest it didn't become apparently clear to me what was meant exactly:

> `no-floating-promises` only detects unhandled Promise _statements_.

---

> The hint about stronger types is to secure your function inputs so that TS can catch when you're passing promises into places they may not be correctly handled.
> `any/unknown` should be a last resort!

I understand and agree, as I mentioned the `any` type was only there for demonstration purposes, it's not a production code or anything like that. Given the tips you just gave me, this whole thing can be re-written as:

```typescript
export const myFooFunc = async (value: Awaited<Promise<string>>): Promise<string> => {
    return new Promise((resolve) => resolve('foo'));
};

export const myBarFunc = async (value: string): Promise<string> => {
    return new Promise((resolve) => resolve('bar'));
};

/** If no-unused-vars, this will be marked as a problem. */
const promise = myFooFunc(await myBarFunc('hello'));

/** Otherwise: */

/** ✅ Marked as a problem. */
promise;

/** ✅ Works as intended */
await promise;
```
